_posts/2023-05-23-GUIs-in-PyQt5.md

```diff
@@ -0,0 +1,263 @@
+        the same as in C++
+      - Providing PEP 484 type hinting for all APIs
+  - <https://www.riverbankcomputing.com/static/Docs/PyQt5/index.html>
+
+## PyQt: Python Documentation
+
+  - Most documentation written by third parties
+  - Easiest to read C++ documentation and hand-translate examples to
+    Python.
+  - Python Wiki: <https://wiki.python.org/moin/PyQt/>
+  - Book: Michael Herrmann, "*Python and Qt: the best parts*"
+  - *(I was able to learn from free online resources.)*
+
+## Building blocks (1): widges, events, properties
+
+  - **Widgets:** tab bars, scroll bars, push buttons, list views,
+    pull-down menus
+  - **Event handlers:** callback functions responding to *primitive*
+    mouse/keyboard events.
+  - **Properties:** fields with getters and setters, changes
+    automatically updates widget drawing on screen
+      - e.g. a text fields current string
+
+## Building blocks (2): signals, slots, connections
+
+  - **Signals:** complex combination of primitive events that trigger
+    widget-specific actions
+      - e.g. when list view item is double-clicked
+  - **Slots:** public methods used as callbacks, can be "connected" to
+    signals to be triggered on certain events.
+  - **Connections:** are an object that maintain a link between a signal
+    and slot. Can be disconnected/deleted
+  - Using these building blocks, simple interfaces can be designed
+    without the need for custom widgets.
+
+## Object oriented programming: method overriding
+
+  - When you want to combine multiple widgets into a single more complex
+    widget.
+  - Signals and slots are usually not enough.
+  - A custom widget inherits from a parent widget (e.g. `QWidget` or
+    `QListView`)
+  - Override signals and slots to customize behavior (e.g. update state
+    variables)
+
+<https://doc.qt.io/qt-5/qtwidgets-index.html>
+
+## Declaring your own reusable widget
+
+  - Most code occurs in `__init__()` constructor
+  - Create sub-widgets and layout
+  - Place sub-widgets into layout object:
+      - `QBoxLayout`
+      - `QFormLayout`
+      - `QGridLayout`
+      - `QStackedLayout`
+  - Define slots to update model: `@pyqtSlot()` decorator
+  - Connect child widget signals to slots
+
+## The **Graphics View Framework**
+
+  - Interactive 2D visualizations
+  - Arbitrarily sized canvas (`QGraphicsScene`)
+  - Easily renders large number of objects (`QGraphicsItem`)
+  - Graphics view widget provides (`QGraphicsView`)
+      - translating (with scroll bars)
+      - scaling
+      - rotation
+
+<https://doc.qt.io/qt-6/graphicsview.html>
+
+## Diagram Scene Example (tutorial)
+
+![](../../../assets/images/2023-05-23-GUIs-in-PyQt5-pics/diagramscene.png)
+
+<https://doc.qt.io/qt-5/qtwidgets-graphicsview-diagramscene-example.html>
+
+## Pattern Matching Tool Demo
+
+I created an app to run a pattern matching algorithm using OpenCV to
+select parts of a "target" image that are similar to a "pattern" image.
```

Comment:
```suggestion
## Pattern Matching Tool

I created an app to run a pattern matching algorithm using OpenCV to
select parts of a "target" image that are similar to a "pattern" image.
```
We can have some screenshots here to show how it looks like. This is quite important information for readers.
Is D&D realized? Was that the standard already implemented in a widget? You can mention if you did something additionally.
You can also mention, which new features might be nice for users. Points of possible improvement.

How about Image cropping tool?